### ISSUE TITLE
How to make NO_DHCP work?

### ISSUE THREAD

**Issue by barraudf** [rancher/old-vm#44]
Hi,

I have a dhcp server on the host's network, which works fine for physical machines.

I'd like my rancher VMs to have their network set by this dhcp server.

As documented, I tried to run a vm with :

> docker run -d -e "RANCHER_VM=true" --cap-add NET_ADMIN -v     /tmp/ranchervm:/ranchervm --device /dev/kvm:/dev/kvm -p 22   --device /dev/net/tun:/dev/net/tun -e "NO_DHCP=true" rancher/vm-ubuntu -m 1024 -smp 1 -k fr

But the vm won't get any IP.
(Running with the container internal dnsmasq works fine, but I'd like to get rid of it)

Did I miss anything?

**Comment by sheng-liang** [comment]
The VM would be issuing DHCP requests on the Docker bridge, which is NAT'ed to the host network by the Docker daemon using iptables.

So you either need to somehow forward DHCP requests over NAT or run a DHCP server on Docker bridge.

RancherVM by default runs a DHCP server for each VM on the Docker bridge.

**Comment by barraudf** [comment]
Ok, this may not be releated to ranchervm, I can't get dhcp to work in any of my containers. My guess is the problem is between docker0 and container's eth0 but I'm not a network expert.

I got dhcp working using a second physical NIC and pipework. I'll try this way. What I want in the end is a rancherVM having its eth0 connected to my physical local network.

**Comment by sheng-liang** [comment]
It is possible to configure RancherVM to bridge its eth0 to host network.

Assuming you are running Ubuntu host and your host is connected to its physical network via `eth0`.

First you need to create a bridge in your host for `eth0`. Setup your `/etc/network/interfaces` file as follows:

```
auto lo
iface lo inet loopback

auto eth0
iface eth0 inet manual

auto br0
iface br0 inet dhcp
    bridge_ports eth0
    bridge_stp off
    bridge_fd 0
    bridge_maxwait 0
```

Running ifconfig on the host should give you something like this:

```
# ifconfig
br0       Link encap:Ethernet  HWaddr 00:0c:29:24:8a:81  
      inet addr:192.168.111.158  Bcast:192.168.111.255  Mask:255.255.255.0
      inet6 addr: fe80::20c:29ff:fe24:8a81/64 Scope:Link
      UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
      RX packets:2354 errors:0 dropped:0 overruns:0 frame:0
      TX packets:2012 errors:0 dropped:0 overruns:0 carrier:0
      collisions:0 txqueuelen:0 
      RX bytes:1202270 (1.2 MB)  TX bytes:250767 (250.7 KB)

eth0      Link encap:Ethernet  HWaddr 00:0c:29:24:8a:81  
      UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
      RX packets:1539 errors:0 dropped:0 overruns:0 frame:0
      TX packets:1347 errors:0 dropped:0 overruns:0 carrier:0
      collisions:0 txqueuelen:1000 
      RX bytes:580472 (580.4 KB)  TX bytes:209948 (209.9 KB)
```

Note IP address is now set on `br0`. And `eth0` and `br0` have the same MAC.

You can now change Docker daemon to use `br0` instead of `docker0`:

```
# service docker stop
# echo 'DOCKER_OPTS="-b=br0"' >> /etc/default/docker
# service docker start
```

Now RancherVMs started with `NO_DHCP=true` will get their IP from host DHCP.

One gotcha: Docker will assign IPs from the same range to regular non-RancherVM containers. To avoid IP conflcits, you may want to provide an IP range for Docker containers to the Docker daemon `--fixed-cidr=...`

**Comment by barraudf** [comment]
I tried something like this (except that I bridged eth1 so this didn't mess with docker networking), but I couldn't get network to work in VM, even with a static address.

But I finally got it working, manually bridging `eth1` to `br0` on host, running a container with `--net=none` and creating its network with pipework. Now the vm is seen on the network as any other physical machine.

Something wierd is that the container and the vm has the same IP, but as long as it works, I don't mind.

Thank you!
